ancestry.go

```diff
@@ -0,0 +1,283 @@
+		}
+	}
+}
+
+func (b *AncestryBuilder) createLayerIndexedFeature(namespace *layerIndexedNamespace, feature *database.LayerFeature) layerIndexedFeature {
+	return layerIndexedFeature{
+		feature:      feature,
+		namespace:    namespace,
+		introducedIn: b.layerIndex,
+	}
+}
+
+func (b *AncestryBuilder) lookupNamespace(feature *database.LayerFeature) (*layerIndexedNamespace, bool) {
+	for _, namespace := range b.namespaces {
+		if namespace.namespace.VersionFormat == feature.VersionFormat {
+			return namespace, true
+		}
+	}
+
+	return nil, false
+}
+
+func (b *AncestryBuilder) ancestryFeatures(index int) []database.AncestryFeature {
+	ancestryFeatures := []database.AncestryFeature{}
+	for detector, features := range b.features {
+		for _, feature := range features {
+			if feature.introducedIn == index {
+				ancestryFeatures = append(ancestryFeatures, database.AncestryFeature{
+					NamespacedFeature: database.NamespacedFeature{
+						Feature:   feature.feature.Feature,
+						Namespace: feature.namespace.namespace.Namespace,
+					},
+					FeatureBy:   detector,
+					NamespaceBy: feature.namespace.namespace.By,
+				})
+			}
+		}
+	}
+
+	return ancestryFeatures
+}
+
+func (b *AncestryBuilder) ancestryLayers() []database.AncestryLayer {
+	layers := make([]database.AncestryLayer, 0, b.layerIndex)
+	for i := 0; i < b.layerIndex; i++ {
+		layers = append(layers, database.AncestryLayer{
+			Hash:     b.layerNames[i],
+			Features: b.ancestryFeatures(i),
+		})
+	}
+
+	return layers
+}
+
+// Ancestry produces an Ancestry from the builder.
+func (b *AncestryBuilder) Ancestry(name string) *database.Ancestry {
+	if name == "" {
+		// TODO(sidac): we'll use the computed ancestry name in the future.
+		// During the transition, it still requires the user to use the correct
+		// ancestry name.
+		name = ancestryName(b.layerNames)
+		log.WithField("ancestry.Name", name).Warn("generated ancestry name since it's not specified")
+	}
+
+	return &database.Ancestry{
+		Name:   name,
+		By:     b.detectors,
+		Layers: b.ancestryLayers(),
+	}
+}
+
+// SaveAncestry saves an ancestry to the datastore.
+func SaveAncestry(store database.Datastore, ancestry *database.Ancestry) error {
+	log.WithField("ancestry.Name", ancestry.Name).Debug("saving ancestry")
+	if err := database.UpsertAncestryAndCommit(store, ancestry); err != nil {
+		log.WithError(err).WithFields(log.Fields{"ancestry.Name": ancestry.Name}).Error("failed to save ancestry to datastore")
+		serialized, _ := json.Marshal(ancestry)
+		log.Debug(string(serialized))
+		return StorageError
+	}
+
+	return nil
+}
+
+// IsAncestryCached checks if the ancestry is already cached in the database with the current set of detectors.
+func IsAncestryCached(store database.Datastore, name string, layerHashes []string) (bool, error) {
+	if name == "" {
+		// TODO(sidac): we'll use the computed ancestry name in the future.
+		// During the transition, it still requires the user to use the correct
+		// ancestry name.
```

Comment:
Because when getting the ancestry, it requires a name.
We can either do: 
return the computed name to the client
or
the client still gives us the name, but we use an extra table to keep the client given name to the internal ancestry name mapping.

Reply:
I think we should get rid of names and make the computed name an implementation detail. Just let the customer provide the layer list.

Reply:
OR (which I prefer)
The user always query with a list of layer hashes in the request body. This just requires us to change the API a little bit.

Reply:
exactly